text/0018-bee-bundle/0018-bee-bundle.md

```diff
@@ -0,0 +1,281 @@
+| | transaction.trunk = trunk
+| | transaction.branch = branch
+| | if transaction.current_index == transaction.last_index
+| | | branch = trunk
+| | transaction.attachment_timestamp = timestamp()
+| | transaction.attachment_timestamp_lower = 0
+| | transaction.attachment_timestamp_upper = 3812798742493
+| | if transaction.tag.empty()
+| | | transaction.tag = transaction.obsolete_tag
+| | trunk = transaction.pow(mwm)
+
+```
+
+### Validate
+
+*Client side and server side operation.*
+
+Validating a bundle means checking the syntactic and semantic integrity of a bundle as a whole and of its constituent transactions. As bundles are atomic transfers, either all or none of the transactions will be accepted by the network. After validation, transactions of a bundle are candidates to be included to the ledger.
+
+For a bundle to be considered valid, the following assertions must be true:
+
+- bundle has announced size;
+- transactions share the same bundle hash;
+- transactions absolute value doesn't exceed total IOTA supply;
+- bundle absolute sum never exceeds total IOTA supply;
+- order of transactions in the bundle is the same as announced by `current_index` and `last_index`;
+- value transactions have an address ending in `0` i.e. has been generated with Kerl;
+- bundle inputs and outputs are balanced i.e. the bundle sum equals `0`;
+- announced bundle hash matches the computed bundle hash;
+- for spending transactions, the signature is valid;
+
+Pseudocode:
+
+```
+validate(bundle):
+| value = 0
+| current_index = 0
+|
+| if bundle.length() != bundle.at(0).last_index + 1
+| | return BUNDLE_INVALID_LENGTH
+|
+| bundle_hash = bundle.at(0).bundle_hash
+| last_index = bundle.at(0).last_index
+|
+| for transaction in bundle
+| | if transaction.bundle_hash != bundle_hash
+| | | return BUNDLE_INVALID_HASH
+| | if abs(transaction.value) > IOTA_SUPPLY
+| | | return BUNDLE_INVALID_TRANSACTION_VALUE
+| | value = value + transaction.value
+| | if abs(value) > IOTA_SUPPLY
+| | | return BUNDLE_INVALID_VALUE
+| | if transaction.current_index != current_index++
+| | | return BUNDLE_INVALID_INDEX
+| | if transaction.last_index != last_index
+| | | return BUNDLE_INVALID_INDEX
+| | if transaction.value != 0 && transaction.address.last != 0
+| | | return BUNDLE_INVALID_ADDRESS
+|
+| if value != 0
+| | return BUNDLE_INVALID_VALUE
+| if bundle_hash != bundleHash(bundle)
+| | return BUNDLE_INVALID_HASH
+| if !bundleSignaturesValidate(bundle)
+| | return BUNDLE_INVALID_SIGNATURE
+|
+| return BUNDLE_VALID
+```
+
+# Drawbacks
+
+<!-- TODO -->
+
+# Rationale and alternatives
+
+- A `Bundle` is a fundamental component of the IOTA protocol and must be implemented;
+- There is no more intuitive and simple way to implement a `Bundle` than the one proposed;
+- Since bundles are final, `BundleBuilder` is mandatory;
+
+# Unresolved questions
+
+- Should this RFC expands a bit more on the M-Bug ? Or give a link ?
+- Should `Bundle` provide a `.transactions` or a `.at` method ?
```

Comment:
Is there a natural order to `Bundle`? Can `Transactions` arrive out of order?

The `.at` method is basically indexing into the `Bundle`, getting a specific transaction? I think this makes sense, if there is a natural order.

Reply:
Yes there is, `.at()` it is then.